### PR TITLE
minor cleanup for Model rendering code

### PR DESCRIPTION
**Pull request by AndrewMeadows**

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 87bdaae05acc2aa9fe82e71d80d1a183cbaa9846 available at: https://gist.github.com/24c0906d2a1a8d8d8f2f150c6fc95163<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9358-87bdaae.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9358-87bdaae.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 87bdaae05acc2aa9fe82e71d80d1a183cbaa9846 available at: https://gist.github.com/24c0906d2a1a8d8d8f2f150c6fc95163<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9358-87bdaae.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9358-87bdaae.dmg

**Comment by AndrewMeadows** [comment]
test this please

File: libraries/render-utils/src/MeshPartPayload.cpp [hunk 1]
```diff
@@ -63,8 +63,7 @@ void MeshPartPayload::updateMeshPart(const std::shared_ptr<const model::Mesh>& d
 
 void MeshPartPayload::updateTransform(const Transform& transform, const Transform& offsetTransform) {
     _transform = transform;
-    _offsetTransform = offsetTransform;
-    Transform::mult(_drawTransform, _transform, _offsetTransform);
+    Transform::mult(_drawTransform, _transform, offsetTransform);
```
Comment (AndrewMeadows) [inline libraries/render-utils/src/MeshPartPayload.cpp]: `MeshPartPayload::_offsetTransform` was cruft: it was set here but never used elsewhere.

File: libraries/render-utils/src/MeshPartPayload.cpp [hunk 2]
```diff
@@ -371,8 +370,10 @@ void ModelMeshPartPayload::updateTransformForSkinnedMesh(const Transform& transf
             _worldBound += clusterBound;
         }
 
-        // clusterMatrix has world rotation but not world translation.
-        _worldBound.translate(transform.getTranslation());
+        _worldBound.transform(transform);
+        if (clusterMatrices.size() == 1) {
+            _transform = _transform.worldTransform(Transform(clusterMatrices[0]));
+        }
```
Comment (AndrewMeadows) [inline libraries/render-utils/src/MeshPartPayload.cpp]: Non-skinned models have only one clusterMatrix (e.g. only one "joint") which transforms from "model" to "world" so we properly update `_transform` from it.  This means we can not bother handling this case in `ModelMeshPartPayload::bindTransform()`.

File: libraries/render-utils/src/MeshPartPayload.cpp [hunk 3]
```diff
@@ -520,23 +521,15 @@ void ModelMeshPartPayload::bindTransform(gpu::Batch& batch, const ShapePipeline:
...
         } else {
             batch.setUniformBuffer(ShapePipeline::Slot::BUFFER::SKINNING, state.clusterBuffer);
         }
-    } else {
-        if (canCauterize && _model->getCauterizeBones()) {
-            transform = Transform(state.cauterizedClusterMatrices[0]);
-        } else {
-            transform = Transform(state.clusterMatrices[0]);
-        }
     }
```
Comment (AndrewMeadows) [inline libraries/render-utils/src/MeshPartPayload.cpp]: When `state.clusterBuffer` is null it means there was only one clusterMatrix which means we already stored the correct values in `_transform`, which is why we can remove the `else` case above.

File: libraries/render-utils/src/Model.cpp [hunk 4]
```diff
@@ -92,7 +92,6 @@ Model::Model(RigPointer rig, QObject* parent) :
     _snapModelToRegistrationPoint(false),
     _snappedToRegistrationPoint(false),
     _cauterizeBones(false),
-    _pupilDilation(0.0f),
```
Comment (AndrewMeadows) [inline libraries/render-utils/src/Model.cpp]: `Model::_pupilDilation` is ancient cruft from the days when the only `Model` instances were for avatars.

File: libraries/render-utils/src/Model.cpp [hunk 5]
```diff
@@ -1167,28 +1172,26 @@ void Model::updateClusterMatrices(glm::vec3 modelPosition, glm::quat modelOrient
...
-            glm_mat4_mul((glm_vec4*)&modelToWorld, (glm_vec4*)&jointMatrix, (glm_vec4*)&temp);
-            glm_mat4_mul((glm_vec4*)&temp, (glm_vec4*)&inverseBindMatrix, (glm_vec4*)&out);
+            glm::mat4 out, inverseBindMatrix = cluster.inverseBindMatrix;
+            glm_mat4_mul((glm_vec4*)&jointMatrix, (glm_vec4*)&inverseBindMatrix, (glm_vec4*)&out);
             state.clusterMatrices[j] = out;
-#else 
-            state.clusterMatrices[j] = modelToWorld * jointMatrix * cluster.inverseBindMatrix;
+#else
+            state.clusterMatrices[j] = jointMatrix * cluster.inverseBindMatrix;
 #endif
```
Comment (AndrewMeadows) [inline libraries/render-utils/src/Model.cpp]: This is where we expect to see a small optimization: we don't bother multiplying everything by `modelToWorld` (which used to only include the rotation part).  Instead we use the proper `_transform` at the end of `ModelMeshPartPayload::bindTransform()`.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 87bdaae05acc2aa9fe82e71d80d1a183cbaa9846 available at: https://gist.github.com/24c0906d2a1a8d8d8f2f150c6fc95163<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9358-87bdaae.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9358-87bdaae.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: e2130d33d9e70eb7f892715db0e4bc07fe0cf72f available at: https://gist.github.com/24c0906d2a1a8d8d8f2f150c6fc95163<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9358-e2130d3.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9358-e2130d3.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: b3e336de3dd0d2e696eff0b7aa63b30fd4d3e4d4 available at: https://gist.github.com/24c0906d2a1a8d8d8f2f150c6fc95163<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9358-b3e336d.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9358-b3e336d.dmg

**Comment by AndrewMeadows** [comment]
The only things that have changed are how transforms of skinned models (avatars and animated entities) are computed for the render pipeline.  There are no new features or bug fixes.  Theoretically this PR's frame rate should be faster or more regular than before, but the effects are probably not measurable without going into a region with 100 simultaneous avatars.

So, what to test:

(1) Login to a domain with at least one other avatar --> the other avatar should render and animate as normal.

(2) Login to a domain that has an animated entity (dev-playa could work in a pinch: it has an animated robot on top of the "Burgerdrome" sign near <xyz> = <-34, -8, -21>).  The animated entity should still move around correctly.

(3) ... TODO: finish this test plan

**Comment by hifi-gustavo** [comment]
Build console output for commit: c060fd6a68b3d669cca61d050d2bfd1a07f0998a available at: https://gist.github.com/24c0906d2a1a8d8d8f2f150c6fc95163<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9358-c060fd6.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9358-c060fd6.dmg

**Comment by nbq** [comment]
I will test this one

**Comment by nbq** [comment]
Test done by Coal in Desktop Mode:

(1) PASS, went to Welcome and everyone showed up and animated as normal

(2) PASS, went to dev-playa and could see the animating entity on the "Burgerdome" sign just fine

(3) PASS, Went to `dev-chris.highfidelity.io` first to load up on the cache - on a cold cache the frame rate drops down to 12 from the usual 59-60, then restarted interface for the "warm cache" - logged back into `dev-chris.highfidelity.io` and noticed the frame rate is still around 12 though it did load a lot faster. *I say PASS because I tested with the newest RC and the frame rate is the same as in the test PR*

In reply to NOTE: I saw avatars rendered on top of me on the RC when I tested but not on the PR, so I wonder if there is some randomness to it since we saw it opposite each other.

**Comment by cozza13** [comment]
Hi @claudioshg when you went to dev-chris.highfidelity.io did you see a lot of avatars standing around?

**Comment by AndrewMeadows** [comment]
This is a 10-second trace of the current Release in dev-chris domain starting 30 seconds after login.  I only selected the first five seconds of the trace because in both the Release and PR-9358 cases the interface started to "reduce rendered detail" in the last few seconds of the trace.
trace-release-main-thread-20170112

This is the corresponding trace for PR-9358:
trace-pr9358-main-thread-20170112

Pay attention to the Average Wall Duration times.  The good news is that they go down (down is better!) a little bit pretty much across the board in this PR.

**Comment by claudioshg** [comment]
@cozza13 Yes all 100 avatars loaded. 
*I faced them with my HMD, in first person view, and took note at the frame rate. It was 90fps. It deviated from 90fps ~.06 fps.

**Pull request merged**